x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java

```diff
@@ -299,6 +299,7 @@ static void persistProgress(Client client, String jobId, Runnable runnable) {
                 }
                 IndexRequest indexRequest = new IndexRequest(indexOrAlias)
                     .id(progressDocId)
+                    .setRequireAlias(AnomalyDetectorsIndex.jobStateIndexWriteAlias().equals(indexOrAlias))
```

Comment:
One concern I have about this pattern of figuring out the value of `requireAlias` is that we need to be very careful to provide it with the right value as there is nothing (other than reading the code ;)) that could help us here.
Ideally, I would like to see `AnomalyDetectorsIndex.jobStateIndexWriteAlias()` (and other methods of this kind) return a tuple (`AliasOrIndex`) with two fields: `name` and `isAlias`. This way the two pieces of information would be bundled together.
However, this may be a big code change and the benefits are not easy to measure.

Alternatively, you could have a local variable `boolean requireAlias` initialized to `true` in line 296 and set to `false` in line 298. This way, the decisions whether to use alias or index and whether to require alias would be logically grouped.

Reply:
Note that `AnomalyDetectorsIndex.jobStateIndexWriteAlias()` always returns the alias. The issue here is that if we detect existing docs we want to overwrite them in the same index they exist rather than where the alias currently points at.

Reply:
I think the core issue here is the "index or alias" check, not what our original methods return. Callers of the method know they are getting an alias and should handle it accordingly. 

Here, we don't know because we searched for the concrete index name as well.

Reply:
Sure, the `jobStateIndexWriteAlias` method cannot know that.
I was just thinking out loud whether some additional encapsulation would make the code cleaner but I don't see an easy way to fix it.